File: reducing-dishes.py
class Solution:
    def maxSatisfaction(self, satisfaction: List[int]) -> int:
        positiveSatisfaction = 0
        posSatisfaction = []
        negSatisfaction = []
        time = 1
        N = len(satisfaction)
        ans = 0

        for i in range(N):
            if satisfaction[i] >= 0:
                posSatisfaction.append(satisfaction[i])
            else:
                negSatisfaction.append(satisfaction[i])
        
        posSatisfaction.sort()
        negSatisfaction.sort(reverse=True)

        for i in range(len(posSatisfaction)):
            if posSatisfaction[i] >= 0:
                ans += time*posSatisfaction[i]
                positiveSatisfaction += posSatisfaction[i]
                time += 1
        
        for i in range(len(negSatisfaction)):
            if negSatisfaction[i] < 0 and -negSatisfaction[i] < positiveSatisfaction:
                positiveSatisfaction += negSatisfaction[i]
                ans += positiveSatisfaction
            else:
                break

        return ans

File: reducing-dishes.cpp
class Solution {
public:
    int maxSatisfaction(vector<int>& satisfaction) {
        sort(satisfaction.begin(), satisfaction.end());
        int n = satisfaction.size();
        int dp[n+1][n+1];

        memset(dp, 0, sizeof dp);
        for(int i = 0; i < n; i ++)dp[i][0] = 0;

        for(int j = 1; j <= n; j++){
            for(int i = 1; i <= n; i++){
                dp[i][j] = satisfaction[i-1]*j + dp[i-1][j-1];
            }
        }

        int ans = 0;
        for(int i = 0; i <= n; i++)ans = max(ans, dp[n][i]);
        return ans;
    }
};